[MERGE] base: remove field source of ir.translation

mart-e opened the pull request:
Instead, make sure the field src is always up to date

Change a bit the behaviour of `_write` to always try to update potential
translations when in monolanguage (was only done in multi-language if different
than en_US) and always create translations when writing on a translatable field
when in multi-language (was skipped in en_US)

Add tests

Task-id: 2031752
Pad: https://pad.odoo.com/p/r.dbf835d0aae979d858f73446dcbfb1df

xmo-odoo commented on odoo/addons/base/models/ir_translation.py:
```diff
@@ -329,6 +278,23 @@ def _set_ids(self, name, tt, lang, ids, value, src=None):
+        """ Update the translation source of records.
+
+        :param name: a string defined as "<model_name>,<field_name>"
+        :param tt: the type of translation (should always be "model")
```
If this is going to be assumed, shouldn't we enforce it?

xmo-odoo replied:
Yeah but I think it could be modified at that point.

xmo-odoo commented on odoo/addons/base/tests/test_translate.py:
```diff
@@ -328,6 +342,57 @@ def test_105_duplicated_translation(self):
+
+        # Translate in both language
+        translations[0].write({'value': 'The Cheese'})
+        translations[1].write({'value': 'Fromage'})
```
Since these lines are updating one record each, maybe use setattr for clarity / simplicity?

xmo-odoo commented on odoo/models.py:
```diff
@@ -3447,9 +3447,9 @@ def _write(self, vals):
                     self.env['ir.translation']._sync_terms_translations(field, self)
 
-                elif has_translation and field.translate:
+                elif self.env.lang and field.translate:
```
Is it intended that when the lang is set to en_US we'll both change the base term (line 3409) and set a translation?

mart-e replied:
Yes, this is more a bugfix IMHO. In case you have an English translation, making a write was only writing on the master record but the translation was not updated. That was revealed by 73a7534bfc0e098a but could potentially be reproduced in stable so could be worth backporting.

xmo-odoo commented on odoo/models.py:
```diff
@@ -3458,6 +3458,8 @@ def _write(self, vals):
                     self.env['ir.translation']._set_ids(
                         tname, 'model', self.env.lang, self.ids, val, src_trans)
+                    self.env['ir.translation']._set_source(
+                        tname, 'model', self.ids, src_trans)
```
Wouldn't it make sense to update _set_ids to handle the _set_source use case instead? Also possibly rename it because _set_ids is an awful name for it.

mart-e replied:
The main difference is that `_set_ids`  (as the name obviously indicated) is only working on one record but it could be changed I guess (just check `l10n_multilang` that calls this method too.

xmo-odoo replied:
But as far as I can see this is the only call to _set_source and it immediately follows a _set_ids call, so there's effectively no difference?

xmo-odoo commented on odoo/models.py:
```diff
@@ -4283,11 +4285,12 @@ def get_trans(field, old, new):
                     del vals['module']      # duplicated vals is not linked to any module
                     vals['res_id'] = target_id
+                    if not callable(field.translate):
+                        vals['src'] = new_wo_lang[name]
```
Why?

mart-e replied:
Because it was empty before, we used to set `src` only for the `vals['lang'] == old.env.lang` scenario but it must be set for all languages.

mart-e commented:
Ok when trying to split and formalise the last behaviour with a test, I realised it did not worked :)

So I have split the commit and added a new longer test but it changes a bit the logic (so probably not in stable):

If one language only (en_US or not), when writing on a translatable field:
- record field should updated
- value of potential existing en_US translation should be updated
- src of potential existing translations should be updated
- no new translation should be created

If en_US and fr_FR, when writing on a translatable field in en_US:
- record field should updated
- value of potential existing en_US translation should be updated
- src of potential existing translations should be updated
- **new en_US translation should be created if was not present**

If en_US and fr_FR, when writing on a translatable field in fr_FR:
- record field should not be updated
- value of potential existing fr_FR translation should be updated
- src of potential existing translations should not be updated
- new fr_FR translation should be created if was not present

mart-e commented:
@robodoo r+ jacta est ! rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

robodoo commented:
Merged at ed7b53bb1def9d898a766076df3a44e3ac7deeda, thanks!
